### PR TITLE
[HOP] Reworked DispatchKey.Autograd

### PR DESCRIPTION
**Pull request by bohnstingl**
This PR intends to rework the dispatching of the autograd key.
I.e., currently the DispatchKey.Autograd of the HOPs was triggered, even if non of the operands of the HOP have `requires_grad=True`. With this rework, the autograd is bypassed if non of the operands require gradients and only invoked if any of the operands require gradients.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @ydwu4

### CODE REVIEW

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 1]
```diff
@@ -447,7 +447,7 @@ def backward(
         return None, None, None, *grads
 
 
-@invoke_subgraph.py_impl(DispatchKey.Autograd)
+@invoke_subgraph.py_autograd_impl
```
Comment (zou3519) [inline torch/_higher_order_ops/invoke_subgraph.py:450]: this seems redundant for invoke_subgraph

Comment (bohnstingl) [reply torch/_higher_order_ops/invoke_subgraph.py:450]: Thank you for the quick feedback. I think in the case of the `invoke_subgraph` we can remove the checks in `invoke_subgraph.py` and only have them in `_ops.py`. WDYT?

Comment (zou3519) [reply torch/_higher_order_ops/invoke_subgraph.py:450]: yes

**Comment by ydwu4** [comment]
Looks good! How aobut we also replace the pyimpl on those hops which don't have autograd implementation? so that there's no confusion which decorator to use to register an autograd key for hops.

**Comment by bohnstingl** [comment]
@ydwu4 I made an update to also rework the HOPs that do not yet have an autograd implementation.
One testcase of `map` needed some special treatment because of the wrapper that map currently uses. WDYT?

File: torch/_higher_order_ops/flex_attention.py [hunk 2]
```diff
@@ -709,7 +709,7 @@ def backward(ctx: Any, grad_out: Tensor, grad_logsumexp: Tensor) -> tuple[Option
         return grad_query, grad_key, grad_value, *none_grads, *grad_score_mod_captured
 
 
-@flex_attention.py_impl(DispatchKey.Autograd)
+@flex_attention.py_autograd_impl
```
Comment (ydwu4) [inline torch/_higher_order_ops/flex_attention.py:712]: It seems flex_attention have specialized code path when no inputs require grad. Ideally, we should also change the body to be something like:
```python
with TransformGetItemToIndex():
          out, logsumexp = FlexAttentionAutogradOp.apply(
            query,
            key,
            value,
            fw_graph,
            bw_graph,
            block_mask,
            scale,
            kernel_options,
            mask_mod_other_buffers,
            *score_mod_other_buffers,
        )
    return out, logsumexp
```
But it's kind of risky. Let's keep flex_attention unchanged (still using py_impl) in this PR and submit a seperate PR with a test for above proposed change.

**Review by ydwu4 — APPROVED** [review]

File: torch/_higher_order_ops/flex_attention.py (hunk 2 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/flex_attention.py]: Good catch. I have reverted the implementation of the `DispatchKey.Autograd` for flex_attention and left a TODO for a follow-up PR.
